src/proto/h1/mod.rs

```diff
@@ -93,3 +93,26 @@ impl Wants {
         (self.0 & other.0) == other.0
     }
 }
+
+fn parse_headers<T>(bytes: &mut BytesMut, ctx: ParseContext<'_>) -> ParseResult<T::Incoming>
+where
+    T: Http1Transaction,
+{
+    // If the buffer is empty, don't bother entering the span, it's just noise.
+    if bytes.is_empty() {
+        return Ok(None);
+    }
+
+    let span = trace_span!("parse_headers");
+    let _s = span.enter();
+    T::parse(bytes, ctx)
+}
+
+fn encode_headers<T>(enc: Encode<'_, T::Outgoing>, dst: &mut Vec<u8>) -> crate::Result<Encoder>
+where
+    T: Http1Transaction,
+{
+    let span = trace_span!("encode_headers");
```

Comment:
TIOLI: we could mimic the existing `{role=server}`/`{role=client}` logging in a more structured way like this:
```suggestion
    let span = trace_span!("parse_headers", role = %std::any::type_name::<T>());
    let _s = span.enter();
    T::parse(bytes, ctx)
}

fn encode_headers<T>(enc: Encode<'_, T::Outgoing>, dst: &mut Vec<u8>) -> crate::Result<Encoder>
where
    T: Http1Transaction,
{
    let span = trace_span!("encode_headers", role = %std::any::type_name::<T>());
```

Reply:
The role stuff was from `log` days. I think having a top level with the role would be the right way, now, ya?

Reply:
Yeah, that would be ideal --- I thought this might be worth adding for now until a higher-level span is added?